### PR TITLE
done chapter 9_Advanced Login

### PR DESCRIPTION
**Pull request by phatthanh113**
image
**Pull này thực hiện**
1.Remember token and digest
2. Login with remembering
3.Forgetting users
4.“Remember me” checkbox

### CODE REVIEW

**Comment by phatthanh113** [comment]
ready

**Comment by phatthanh113** [comment]
ready

File: app/views/sessions/create.html.erb [hunk 1]
```diff
@@ -0,0 +1,2 @@
+<h1>Sessions#create</h1>
```
Comment (vanvtt-0952) [inline app/views/sessions/create.html.erb]: Xóa file ni nè

File: app/views/sessions/destroy.html.erb [hunk 2]
```diff
@@ -0,0 +1,2 @@
+<h1>Sessions#destroy</h1>
```
Comment (vanvtt-0952) [inline app/views/sessions/destroy.html.erb]: Xóa file ni nè

File: app/views/sessions/new.html.erb [hunk 3]
```diff
@@ -10,7 +10,11 @@
       <%= f.label t ".email" %> 
       <%= f.email_field :email, class: "form-control" %> 
       <%= f.label t ".password" %> 
-      <%= f.password_field :password, class: "form-control" %> 
+      <%= f.password_field :password, class: "form-control" %>
+      <%= f.label :remember_me, class: "checkbox inline" do %>
+      <%= f.check_box :remember_me %>
+        <span>Remember me on this computer</span>
```
Comment (vanvtt-0952) [inline app/views/sessions/new.html.erb]: I18n

File: app/helpers/static_pages_helper.rb [hunk 4]
```diff
@@ -0,0 +1,2 @@
+module StaticPagesHelper
```
Comment (vanvtt-0952) [inline app/helpers/static_pages_helper.rb]: Xóa file

**Comment by phatthanh113** [comment]
ready

File: app/models/user.rb [hunk 5]
```diff
@@ -1,11 +1,33 @@
 class User < ApplicationRecord
   VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i.freeze
+  attr_accessor :remember_token
   validates :name, :email, :password, :password_confirmation, presence: true
   validates :name, :password,
             length: {minimum: Settings.user.password.min_length}
-  validates :email, uniqueness: true, format: {with: VALID_EMAIL_REGEX},
+  validates :email, uniqueness: true
+  validates :email, format: {with: VALID_EMAIL_REGEX},
```
Comment (vanvtt-0952) [inline app/models/user.rb]:
```suggestion
  validates :email, uniqueness: true, format: {with: VALID_EMAIL_REGEX},
```

File: app/models/user.rb [hunk 6]
```diff
@@ -1,11 +1,33 @@
...
+    self.remember_token = User.new_token
+    update_attribute :remember_digest, User.digest(remember_token)
+  end
+
+  def authenticated? remember_token
+    BCrypt::Password.new(remember_digest).is_password? remember_token
+  end
+
+  def forget
+    update_attribute :remember_digest, nil
```
Comment (vanvtt-0952) [inline app/models/user.rb]:
```suggestion
    update_column :remember_digest, nil
```

File: app/models/user.rb [hunk 7]
```diff
@@ -1,11 +1,33 @@
...
+
+  class << self
+    def new_token
+      SecureRandom.urlsafe_base64
+    end
+  end
+
+  def remember
+    self.remember_token = User.new_token
+    update_attribute :remember_digest, User.digest(remember_token)
```
Comment (vanvtt-0952) [inline app/models/user.rb]:
```suggestion
    update_column :remember_digest, User.digest(remember_token)
```

**Comment by phatthanh113** [comment]
ready

**Comment by vanvtt-0952** [comment]
merged

**Pull request merged**